dev/release.sh

```diff
@@ -0,0 +1,695 @@
+EOF
+    exit 0
+}
+
+# Set to one of 'major', 'minor', 'alpha', 'beta' or 'final'
+next_method=
+next_method2=
+
+do_branch=false
+warn_branch=false
+
+do_clean=true
+do_upload=true
+do_update=true
+DEBUG=:
+VERBOSE=:
+git_quiet=-q
+
+force=false
+
+do_help=false
+do_manual=false
+
+tagkey=' -s'
+gpgkey=
+
+upload_address=upload@dev.openssl.org
+
+TEMP=$(getopt -l 'alpha,next-beta,beta,final' \
+              -l 'branch' \
+              -l 'no-upload,no-update' \
+              -l 'verbose,debug' \
+              -l 'local-user:' \
+              -l 'force' \
+              -l 'help,manual' \
+              -n release.sh -- - "$@")
+eval set -- "$TEMP"
+while true; do
+    case $1 in
+    --alpha | --beta | --final )
+        next_method=$(echo "x$1" | sed -e 's|^x--||')
+        if [ -z "$next_method2" ]; then
+            next_method2=$next_method
+        fi
+        shift
+        if [ "$next_method" = 'final' ]; then
+            do_branch=true
+        fi
+        ;;
+    --next-beta )
+        next_method2=$(echo "x$1" | sed -e 's|^x--next-||')
+        shift
+        ;;
+    --branch )
+        do_branch=true
+        warn_branch=true
+        shift
+        ;;
+    --no-upload )
+        do_upload=false
+        shift
+        ;;
+    --no-update )
+        do_update=false
+        shift
+        ;;
+    --verbose )
+        VERBOSE=echo
+        git_quiet=
+        shift
+        ;;
+    --debug )
+        DEBUG=echo
+        do_upload=false
+        shift
+        ;;
+    --local-user )
+        shift
+        tagley=" -u $1"
+        gpgkey=" -u $1"
+        shift
+        ;;
+    --force )
+        force=true
+        shift
+        ;;
+    --help )
+        usage
+        exit 0
+        ;;
+    --manual )
+        sed -e '1,/^### BEGIN MANUAL/d' \
+            -e '/^### END MANUAL/,$d' \
+            < "$0" \
+            | pod2man \
+            | man -l -
+        exit 0
+        ;;
+    -- )
+        shift
+        break
+        ;;
+    * )
+        echo >&2 "Unknown option $1"
+        shift
+        exit 1
+        ;;
+    esac
+done
+
+$DEBUG >&2 "DEBUG: \$next_method=$next_method"
+$DEBUG >&2 "DEBUG: \$next_method2=$next_method2"
+
+$DEBUG >&2 "DEBUG: \$do_branch=$do_branch"
+
+$DEBUG >&2 "DEBUG: \$do_upload=$do_upload"
+$DEBUG >&2 "DEBUG: \$do_update=$do_update"
+$DEBUG >&2 "DEBUG: \$DEBUG=$DEBUG"
+$DEBUG >&2 "DEBUG: \$VERBOSE=$VERBOSE"
+$DEBUG >&2 "DEBUG: \$git_quiet=$git_quiet"
+
+case "$next_method+$next_method2" in
+    major+major | minor+minor )
+        # These are expected
+        ;;
+    alpha+alpha | alpha+beta | beta+beta | final+final | + | +beta )
+        # These are expected
+        ;;
+    * )
+        echo >&2 "Internal option error ($next_method, $next_method2)"
+        exit 1
+        ;;
+esac
+
+# Verbosity feed for certain commands
+VERBOSITY_FIFO=/tmp/openssl-$$.fifo
+mkfifo -m 600 $VERBOSITY_FIFO
+( cat $VERBOSITY_FIFO | while read L; do $VERBOSE "> $L"; done ) &
+exec 42>$VERBOSITY_FIFO
```

Comment:
I get this error at this point:

````
dev/release.sh: 178: exec: 42: not found
````

Reply:
What shell are you using?  It clearly doesn't follow POSIX specs in that case...

Reply:
bash

Reply:
See https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_05

Reply:
Also https://pubs.opengroup.org/onlinepubs/009695399/utilities/exec.html

Reply:
Er...  I use bash as well, and famous last words, it works for me.

Reply:
@mattcaswell and I found that /bin/sh is often linked to dash on Debian-based systems, and that dash isn't quite POSIXly correct, so I'm switching the shebang to specify bash.

Reply:
[obligatory note that assuming bash is present is a linux-ism]

Reply:
OS/X has moved from bash to zsh, I wouldn't be entirely surprised if bash disappeared after a couple more releases.

Reply:
I try to write POSIXly correctly. If zsh is POSIXly correct, I don't anticipate problems.